### ISSUE TITLE
BackHandler does not work when the app reopen

### ISSUE THREAD

**Issue by luckcoding** [react-community/react-navigation#3182]
In my app I try to make a BackHandler listener in the root component to handle the back navigation.

The strange thing is that it is working all the time when the app never turn to background, but if the app reopen, and the route stack is not just the root (**if it is only one route stack, the BackHandler is always working**), the BackHandler is not working the first time. If press the android back button, it will dispatch an action `Navigation/NAVIGATE`, just like this:

```
{
	type: 'Navigation/NAVIGATE',
	routeName: 'DrawerClose'
}
```

Then the route stack is disorderly. I can't find some thing where i did wrong. Here is my code:

```
class RootContainer extends Component {
  componentDidMount () {
    if (Platform.OS === 'android') {
      this._BackHandler()
    }
  }

  _BackHandler = () => {
    BackHandler.addEventListener('hardwareBackPress', () => {
      const { routes } = this.props.nav
      if (routes.length > 1) {
        this.props.back({ key: routes[routes.length - 1]['key'] })
        return true
      }
      if (this.lastBackPressed && (this.lastBackPressed + 2000 >= Date.now())) {
        BackHandler.exitApp()
        return true
      }
      this.lastBackPressed = Date.now()
      Toast.show('再按一次退出应用')
      return true
    })
  }

  componentWillUnmount () {
    BackHandler.removeEventListener('hardwareBackPress')
  }

  render () {
    return (
      <View style={{ flex: 1 }}>
        {/* Platform.OS === 'ios' && <StatusBar barStyle='dark-content' /> */}
        <ReduxNavigation />
      </View>
    )
  }
}

const mapStateToProps = state => ({
  nav: state.nav
})

const mapDispatchToProps = dispatch => ({
  startup: () => dispatch(StartupActions.startup()),
  back: (params) => dispatch(NavigationActions.back(params))
})

export default connect(mapStateToProps, mapDispatchToProps)(RootContainer)
```

Forgive me for my poor English.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 1.0.0-beta.19
| react-native     | 0.48.4
| node             |
| npm or yarn      | yarn

**Comment by luckcoding** [comment]
### Add

This is the configuration of the navigation:

```
export const TabNavigation = DrawerNavigator({
  personal: { screen: Personal },
  ...
}, {
  initialRouteName: 'personal',
  contentComponent: props => <SideBar {...props} />
})

export const INITIAL_AUTHEN_ROUTE_NAME = 'Tab'

export const INITIAL_UNAUTHEN_ROUTE_NAME = 'SignIn'

export const APPINTRO_ROUTE_NAME = 'AppIntro'

const GLOBAL_SCREEN = {
  AppIntro: { screen: AppIntro }
}

const AUTHEN_ROUTES = {
  [INITIAL_AUTHEN_ROUTE_NAME]: { screen: TabNavigation },
  setting: { screen: Setting },
  ...
}

const UNAUTHEN_ROUTES = {
  SignIn: { screen: SignIn },
  ...
}

export const AppNavigation = StackNavigator({
  Launcher: { screen: Launcher },
  ...AUTHEN_ROUTES,
  ...UNAUTHEN_ROUTES,
  ...GLOBAL_SCREEN
}, {
  initialRouteName: 'Launcher',
  headerMode: 'none'
})

// handle
const defaultGetStateForAction = AppNavigation.router.getStateForAction
AppNavigation.router.getStateForAction = (action, state) => {
  const { type, routeName } = action

  // twice
  if (state &&
    type === NavigationActions.NAVIGATE &&
    routeName === state.routes[state.routes.length - 1].routeName
  ) return null

  // jump one
  if (state && type === NavigationActions.BACK) {
    const backRoute = state.routes.find(route => route.routeName === action.key)
    if (backRoute) {
      const backRouteIndex = state.routes.indexOf(backRoute)
      const route = {
        ...state,
        routes: state.routes.slice(0, backRouteIndex + 1),
        index: backRouteIndex
      }
      return route
    }
  }

  return defaultGetStateForAction(action, state)
}
```

In Redux:

```
import React from 'react'
import { addNavigationHelpers } from 'react-navigation'
import { connect } from 'react-redux'
import { AppNavigation } from './AppNavigation'

// here is our redux-aware our smart component
function ReduxNavigation ({ dispatch, nav: state, user }) {
  return <AppNavigation navigation={addNavigationHelpers({ dispatch, state })} />
}

const mapStateToProps = ({ user, nav }) => ({ user, nav })

export default connect(mapStateToProps)(ReduxNavigation)
```

**In the sagas, i do some handle.**

The root page is default `Launcher`. When the app is the first be openging, `Launcher` reset to `AppIntro`, when user is not logged in, reset to `SignIn` ...

Here is the code in sagas:

```
export function * redirectFlow () {
  const state = yield select()

  const authenticated = state.user.isLoggedIn
  const app = state.app

  const index = state.nav.index
  const routeName = state.nav.routes[index].routeName

  const redirectRoute = (name, childRouteName) => NavigationActions.reset({
    index: 0,
    actions: [
      NavigationActions.navigate({ routeName: name })
    ]
  })

  if (!app.hasReadAppintro) {
    yield put(redirectRoute(APPINTRO_ROUTE_NAME))
  } else {
    if (authenticated && !isAuthenRouteName(routeName)) {
      let sideRouteName = state.user.role || 'personal'
      yield put(NavigationActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({
            routeName: INITIAL_AUTHEN_ROUTE_NAME,
            action: NavigationActions.navigate({ routeName: sideRouteName })
          })
        ]
      }))
    } else if (!authenticated && !isUnauthenRouteName(routeName)) {
      yield put(redirectRoute(INITIAL_UNAUTHEN_ROUTE_NAME))
    }
  }
}
```

Finally, as I said that the android back button is runing well if the app never turn to the background. I don't know if i described clearly ?

**Comment by luckcoding** [comment]
This problem has been solved. I navigate a new stack in the `DrawerNavigator` contentComponent, and even forget do `DrawerClose`. This code has been worked:
```
jump (routeName) {
    // the route not in DrawerNavigator
    this.props.navigation.navigate('DrawerClose')
    this.props.navigation.navigate(routeName)
  }
```

**Comment by parkmiin** [comment]
In my case, I removed navigationOptions's drawerLockMode and it worked. Hope it helps!

**Comment by denis-shostik** [comment]
@luckcoding I have the same problem, should I just close my drawer?

**Comment by MapYourShow** [comment]
Just wanted to make a note here that for us it was the drawerLockMode value as well. If we remove the lock mode the back button works to close the Drawer if the user leaves the app and reopens.

**Comment by unknown-admin** [comment]
I finally found what to do after more than three hours of internet research and experimentation. All I had to do was turn off the new back callback behaviour that was added to more recent versions of Android so as for the Android OS BackHandler to work properly, especially when reopening the application.

To accomplish this, just include the following in the <activity> tag of the primary start activity in your AndroidManifest.xml:

**_android:enableOnBackInvokedCallback="false_"**


This fixed the problem for me, and even after reopening the application, the BackHandler functions properly.


below is sample example

```
    <application
      android:name=".MainApplication"
      android:label="@string/app_name"
      android:icon="@mipmap/app_logo"
      android:roundIcon="@mipmap/app_logo_rounded"
      android:allowBackup="false"
      android:theme="@style/AppTheme"
      android:enableOnBackInvokedCallback="false"
```
